bin/cheribsdtest/cheribsdtest_vm.c

```diff
@@ -2031,6 +2037,82 @@ CHERIBSDTEST(cheribsdtest_cheri_revoke_lib_fork,
 
 	cheribsdtest_success();
 }
+
+CHERIBSDTEST(revoke_largest_quarantined_reservation,
+    "Verify that the largest quarantined reservation is revoked")
+{
+	const size_t res_size = 0x100000000;
+	void * __capability res;
+	ptraddr_t res_addr;
+	struct procstat *psp;
+	struct kinfo_proc *kipp;
+	struct kinfo_vmentry *kivp;
+	uint pcnt, vmcnt;
+	bool found_res;
+
+	res = CHERIBSDTEST_CHECK_SYSCALL(mmap(NULL, res_size, PROT_READ,
+	     MAP_ANON, -1, 0));
+	res_addr = (ptraddr_t)res;
+	CHERIBSDTEST_CHECK_SYSCALL(munmap(res, res_size));
+
+	psp = procstat_open_sysctl();
+	CHERIBSDTEST_VERIFY(psp != NULL);
+	kipp = procstat_getprocs(psp, KERN_PROC_PID, getpid(), &pcnt);
+	CHERIBSDTEST_VERIFY(kipp != NULL);
+	CHERIBSDTEST_VERIFY(pcnt == 1);
+	kivp = procstat_getvmmap(psp, kipp, &vmcnt);
+	CHERIBSDTEST_VERIFY(kivp != NULL);
+
+	found_res = false;
+	for (u_int i = 0; i < vmcnt; i++) {
+		/*
+		 * Look for an entry containing our reservation.  It
+		 * may have been merged with a previously quarantined
+		 * region so don't expect an exact match.
+		 */
+		if (kivp[i].kve_start <= res_addr &&
+		    kivp[i].kve_end >= res_addr + res_size) {
+			found_res = true;
+			CHERIBSDTEST_VERIFY(kivp[i].kve_type ==
+			    KVME_TYPE_QUARANTINED);
+		}
+	}
+	CHERIBSDTEST_VERIFY2(found_res, "reservation not found in vmmap");
+
+	procstat_freevmmap(psp, kivp);
+
+	/*
+	 * XXX: Assume that the revoker will revoke the largest
+	 * quarantined reservation.
+	 */
+	malloc_revoke();
+
+	kivp = procstat_getvmmap(psp, kipp, &vmcnt);
+	CHERIBSDTEST_VERIFY(kivp != NULL);
+
+	for (u_int i = 0; i < vmcnt; i++) {
+		/*
+		 * Look for an entry containing our reservation.  We
+		 * assuming res_size is large enough that it's the
+		 * reservation we revoke, we shouldn't find it.
+		 *
+		 * XXX: It's possible procstat_getvmmap() could trigger
+		 * reuse of this space, but probalby not since we'll
+		 * have just flushed malloc()'s quarantine list so
+		 * there should be plenty of objects on the free list(s).
```

Comment:
Years and years ago, I hacked something up so that we could get a paged interface to `sysctl_kern_proc_vmmap`; that might be slightly nicer to have, here.  Rebased (but subsequently untested) patches for your consideration up at https://github.com/CTSRD-CHERI/cheribsd/compare/dev...nwf:cheribsd:202212-proc-vmmap-out .